android/src/main/java/com/doublesymmetry/trackplayer/service/MusicService.kt

```diff
@@ -467,6 +474,96 @@ class MusicService : HeadlessJsTaskService() {
 
+    @Suppress("DEPRECATION")
+    fun isForegroundService(): Boolean {
+        val manager = baseContext.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
+        for (service in manager.getRunningServices(Int.MAX_VALUE)) {
+            if (MusicService::class.java.name == service.service.className) {
+                return service.foreground
+            }
+        }
+        Timber.e("isForegroundService found no matching service")
+        return false
+    }
+
+    @MainThread
+    private fun setupForegrounding() {
+        // Implementation based on https://github.com/Automattic/pocket-casts-android/blob/ee8da0c095560ef64a82d3a31464491b8d713104/modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/PlaybackService.kt#L218
+        var notificationId: Int? = null
+        var notification: Notification? = null
+
+        fun startForegroundIfNecessary() {
+            if (isForegroundService()) {
+                Timber.d("skipping foregrounding as the service is already foregrounded")
+                return
+            }
+            if (notification == null) {
+                Timber.d("can't startForeground as the notification is null")
+                return
+            }
+            try {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                    startForeground(
+                        notificationId!!,
+                        notification!!,
+                        ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK
+                    )
+                } else {
+                    startForeground(notificationId!!, notification)
+                }
+                Timber.d("notification has been foregrounded")
+            } catch (error: Exception) {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S &&
+                    error is ForegroundServiceStartNotAllowedException
+                ) {
+                    Timber.e(
+                        "ForegroundServiceStartNotAllowedException: App tried to start a foreground Service when it was not allowed to do so.",
+                        error
+                    )
+                    emit(MusicEvents.PLAYER_ERROR, Bundle().apply {
+                        putString("message", error.message)
+                        putString("code", "android-foreground-service-start-not-allowed")
+                    });
+                }
+            }
+        }
+
+        scope.launch {
+            event.notificationStateChange.collect {
+                when (it) {
+                    is NotificationState.POSTED -> {
+                        Timber.d("notification posted with id=%s, ongoing=%s", it.notificationId, it.ongoing)
+                        notificationId = it.notificationId;
+                        notification = it.notification;
+                    }
+                    else -> {}
+                }
+            }
+        }
+        scope.launch {
```

Comment:
Yeah I expected to be able to do that also, but it seems you really need to wrap them both in a `scope.launch` – otherwise the second one does not run.